readthedocs/doc_builder/backends/sphinx.py

```diff
@@ -390,64 +373,68 @@ def __init__(self, *args, **kwargs):
+    def _post_build(self):
+        """Internal post build to create the ZIP file from the HTML output."""
+        sphinx_build_dir = os.path.join(self.project_path, self.sphinx_build_dir)
+        temp_zip_file = f"/tmp/{self.project.slug}-{self.version.slug}.zip"
         target_file = os.path.join(
-            self.target,
-            '{}.zip'.format(self.project.slug),
+            sphinx_build_dir,
+            f"{self.project.slug}.zip",
         )
-        if not os.path.exists(self.target):
-            os.makedirs(self.target)
-        if os.path.exists(target_file):
-            os.remove(target_file)
-
-        # Create a <slug>.zip file
-        os.chdir(self.old_artifact_path)
-        archive = zipfile.ZipFile(target_file, 'w')
-        for root, __, files in os.walk('.'):
-            for fname in files:
-                to_write = os.path.join(root, fname)
-                archive.write(
-                    filename=to_write,
-                    arcname=os.path.join(
-                        '{}-{}'.format(self.project.slug, self.version.slug),
-                        to_write,
-                    ),
-                )
-        archive.close()
+
+        # FIXME: ``arcname`` is invalid with this command still, but it's close
+        self.run(
+            "zip",
+            "-r",
+            temp_zip_file,
+            self.sphinx_build_dir,
+            cwd=self.project_path,
+            record=False,
+        )
+        self.run("rm", "-r", self.sphinx_build_dir, cwd=self.project_path, record=False)
+        self.run(
+            "mkdir", "-p", self.sphinx_build_dir, cwd=self.project_path, record=False
+        )
+        self.run("mv", temp_zip_file, target_file, cwd=self.project_path, record=False)
 
 
 class EpubBuilder(BaseSphinx):
 
-    type = 'sphinx_epub'
-    sphinx_builder = 'epub'
-    sphinx_build_dir = '_build/epub'
-
-    def move(self, **__):
-        from_globs = glob(os.path.join(self.old_artifact_path, '*.epub'))
-        if not os.path.exists(self.target):
-            os.makedirs(self.target)
-        if from_globs:
-            from_file = from_globs[0]
-            to_file = os.path.join(
-                self.target,
-                '{}.epub'.format(self.project.slug),
+    sphinx_builder = "epub"
+    sphinx_build_dir = "_readthedocs/epub"
+
+    def _post_build(self):
+        """Internal post build to cleanup EPUB output directory and leave only one .epub file."""
+        sphinx_build_dir = os.path.join(self.project_path, self.sphinx_build_dir)
+        temp_epub_file = f"/tmp/{self.project.slug}-{self.version.slug}.epub"
+        target_file = os.path.join(
+            sphinx_build_dir,
+            f"{self.project.slug}.epub",
+        )
+
+        epub_sphinx_filepaths = glob(os.path.join(sphinx_build_dir, "*.epub"))
+        if epub_sphinx_filepaths:
+            # NOTE: we currently support only one .epub per version
+            epub_filepath = epub_sphinx_filepaths[0]
```

Comment:
This seems like another place to raise a warning/error if multiple are generated?

Reply:
Yes and... no 😅 . I decided to not raise an exception here nor at the upload step either because that will change the current behavior. That was basically the main reason to not fail the builds because of this.

This code is mostly copied/adapted from the previous `move()` method that was not failing the build: https://github.com/readthedocs/readthedocs.org/blob/5252be35fb12c4222f56db356af6e92be2fd04c6/readthedocs/doc_builder/backends/sphinx.py#L437-L445

Reply:
I added a warning log to collect what are the projects generating multiple ePubs and be able to communicate to them in the future in case we want to; for example to tell them "We now support multiple ePubs" 👍🏼

Reply:
Cool 👍